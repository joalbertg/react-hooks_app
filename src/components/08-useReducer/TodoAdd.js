import React, { useRef } from 'react';

import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({handleAddTodo}) => {
  const [{ description }, handleInputChange, handleReset] = useForm({
    description: ''
  });

  const inputRef = useRef();

  const handleForm = event => {
    event.preventDefault();

    if(description.trim().length < 2) return;

    const newTask = {
      id: new Date().getTime(),
      desc: description
    }

    handleAddTodo(newTask);
    handleReset();
    // es mala practica llamar al DOM en componentes
    // document.querySelector('input').select();
    inputRef.current.select();
  }

  return(
    <>
      <h4>Add TODO</h4>
      <hr />

      <form
        onSubmit={handleForm}
        className="mt-2"
      >
        <input
          ref={inputRef}
          className="form-control"
          type="text"
          name="description"
          placeholder="Learn ..."
          autoComplete="off"
          autoFocus="on"
          value={description}
          onChange={handleInputChange}
        />
        <button
          type="submit"
          className="btn btn-warning"
        >
          ADD
        </button>
      </form>
    </>
  );
}

