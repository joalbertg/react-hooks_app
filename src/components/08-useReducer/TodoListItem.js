import React from 'react';

export const TodoListItem = ({todo, index, handleToggle, handleRemove}) => {
  return(
    <>
      <li
        className="list-group-item"
        onClick={() => handleToggle(todo.id)}
      >
        <p className={todo.done ? "complete" : ""}>{index + 1}. {todo.desc}</p>
        <button
          className="btn btn-danger"
          onClick={() => handleRemove(todo.id)}
        >
          X
        </button>
      </li>
      <hr />
    </>
  );
}

