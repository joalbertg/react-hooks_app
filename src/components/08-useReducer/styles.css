* {
  box-sizing: border-box;
}

body {
  background-color: black;
  color: white;
  padding: 60px;
}

hr {
  border-top: 1px solid rgba(255, 255, 255, .5);
  margin: 1px 0;
}

.list-group-item {
  background-color: black;
  display: flex;
  justify-content: space-evenly;
  align-items: center;
}

.list-group-item p {
  margin: 0;
  font-size: 1.5rem;
  cursor: pointer;
  flex: 0 1 60%;
}

.list-group-item button {
  flex: 0 1 10%;
}

.complete {
  text-decoration: line-through;
}

form {
  display: flex;
  justify-content: space-between;
}

form input {
  flex: 0 1 80%;
}

