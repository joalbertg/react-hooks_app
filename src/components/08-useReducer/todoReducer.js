const ADD_TODO = 'ADD_TODO';
const TOGGLE_TODO = 'TOGGLE_TODO';
const REMOVE_TODO = 'REMOVE_TODO';

const todoReducer = (state = [], action) => {
  switch(action.type) {
    case ADD_TODO:
      return [...state, { ...action.payload, node: false }];
    case TOGGLE_TODO:
      return state.map(todo => {
        if(todo.id === action.payload.id) {
          todo.done = !todo.done;
        }
        return todo;
      });
    case REMOVE_TODO:
      return state.filter(item => item.id !== action.payload.id);
    default:
      return state;
  }
}

export {
  todoReducer
};

