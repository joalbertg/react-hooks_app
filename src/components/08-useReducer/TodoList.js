import React from 'react';

import { TodoListItem } from './TodoListItem';

export const TodoList = ({todos, handleRemove, handleToggle}) => {
  return(
    todos.map((todo, i) => (
      <TodoListItem
        key={todo.id}
        todo={todo}
        index={i}
        handleToggle={handleToggle}
        handleRemove={handleRemove}
      />
    ))
  );
}

