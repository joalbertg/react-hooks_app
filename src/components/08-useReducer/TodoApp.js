import React, { useReducer, useEffect } from 'react';

import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';
import { todoReducer } from './todoReducer';

import './styles.css';

const init = () => JSON.parse(localStorage.getItem('todos')) ||Â [];

export const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init);

  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos));
  }, [todos]);

  const handleAddTodo = newTask => {
    dispatch({type: 'ADD_TODO', payload: newTask});
  }

  const handleToggle = id => {
    dispatch({type: 'TOGGLE_TODO', payload: {id}});
  }

  const handleRemove = id => {
    dispatch({ type: 'REMOVE_TODO', payload: { id } });
  }

  return(
    <>
      <h1>TodoApp {todos.length}</h1>
      <hr />

      <div className="row mt-5">
        <div className="col-7">
          <ul className="list-group list-group-flush">
            <TodoList
              todos={todos}
              handleToggle={handleToggle}
              handleRemove={handleRemove}
            />
          </ul>
        </div>
        <div className="col-5">
          <TodoAdd handleAddTodo={handleAddTodo} />
        </div>
      </div>
    </>
  );
}

