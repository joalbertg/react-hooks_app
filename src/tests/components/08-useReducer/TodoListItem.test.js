import React from 'react';
import { shallow } from 'enzyme';
import { renderHook, act } from '@testing-library/react-hooks';

import { TodoListItem } from '../../../components/08-useReducer/TodoListItem';
import { demoTodos } from '../../fixtures/demoTodos';

describe('Tests TodoListItem component', () => {
  const index = 1;
  const handleToggle = jest.fn();
  const handleRemove = jest.fn();
  const wrapper = shallow(
    <TodoListItem
      todo={demoTodos[index]}
      index={index}
      handleToggle={handleToggle}
      handleRemove={handleRemove}
    />
  );

  test('Should display correctly', () => {
    const index = 0;
    const wrapper = shallow(<TodoListItem todo={demoTodos[index]} index={index} />);

    expect(wrapper).toMatchSnapshot();
  });

  test('Should have called handleToggle function', () => {
    // click sobre el item demoTodos[index]
    wrapper.find('li').simulate('click', handleToggle);

    expect(wrapper).toMatchSnapshot();
    expect(handleToggle).toHaveBeenCalledWith(demoTodos[index].id);
  });

  test('Should have called handleRemove function', () => {
    // click sobre el item demoTodos[index]
    wrapper.find('button').simulate('click', handleRemove);

    expect(wrapper).toMatchSnapshot();
    expect(handleRemove).toHaveBeenCalledWith(demoTodos[index].id);
  });

  test('Should have complete class if todo.done is true', () => {
    const index = 0;
    const wrapper = shallow(<TodoListItem todo={demoTodos[index]} index={index} />);
    //const pComplete = wrapper.find('.complete');
    const paragraph = wrapper.find('.complete');

    expect(wrapper).toMatchSnapshot();
    expect(paragraph).toBeTruthy();
    expect(paragraph.text()).toBe(`${index + 1}. ${demoTodos[index].desc}`);
    expect(paragraph.hasClass('complete')).toBeTruthy();
  });
});

