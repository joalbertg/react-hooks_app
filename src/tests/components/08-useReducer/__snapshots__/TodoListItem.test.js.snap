// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests TodoListItem component Should display correctly 1`] = `
<Fragment>
  <li
    className="list-group-item"
    onClick={[Function]}
  >
    <p
      className="complete"
    >
      1
      . 
      Learn React
    </p>
    <button
      className="btn btn-danger"
      onClick={[Function]}
    >
      X
    </button>
  </li>
  <hr />
</Fragment>
`;

exports[`Tests TodoListItem component Should have called handleRemove function 1`] = `
<Fragment>
  <li
    className="list-group-item"
    onClick={[Function]}
  >
    <p
      className=""
    >
      2
      . 
      Learn Redux
    </p>
    <button
      className="btn btn-danger"
      onClick={[Function]}
    >
      X
    </button>
  </li>
  <hr />
</Fragment>
`;

exports[`Tests TodoListItem component Should have called handleToggle function 1`] = `
<Fragment>
  <li
    className="list-group-item"
    onClick={[Function]}
  >
    <p
      className=""
    >
      2
      . 
      Learn Redux
    </p>
    <button
      className="btn btn-danger"
      onClick={[Function]}
    >
      X
    </button>
  </li>
  <hr />
</Fragment>
`;

exports[`Tests TodoListItem component Should have complete class if todo.done is true 1`] = `
<Fragment>
  <li
    className="list-group-item"
    onClick={[Function]}
  >
    <p
      className="complete"
    >
      1
      . 
      Learn React
    </p>
    <button
      className="btn btn-danger"
      onClick={[Function]}
    >
      X
    </button>
  </li>
  <hr />
</Fragment>
`;
